# JACKIE.queryDB (bigBedToBedPlus)

JACKIE.queryDB is a modified version of Kent's bigBedToBed which allows the extraction of multiple regions simultaneously, as well as allowing for filtering of integer-based fields of custom bigBed format. Also allows for the selection of a specified number of best bed items according to a series of column sorting criteria.

Graphical User Interface for JACKIE.queryDB now available on Mac as an app

[JACKIE.queryDB.GUI](../JACKIE.queryDB.GUI)

Binaries are available under the JACKIE's GitHub Binaries folder

Architecture | Folder
--- | ---
M1 Mac | [Binaries/arm64-apple-darwin20](../Binaries/arm64-apple-darwin20)
Intel Mac x86_64 | [Binaries/x86_64-Mac](../Binaries/x86_64-Mac)
Linux x86_64 | [Binaries/x86_64-linux](../Binaries/x86_64-linux)



To build from source, skip to [Building JACKIE.queryDB](#building-JACKIE.queryDB)

Usage: JACKIE.queryDB [-filter <filterFile>] input.bb 
For convenient piping, I have decided to change output to STDOUT instead of writing to a file in bigBedToBed. To write to file, use redirection (i.e., JACKIE.queryDB [-filter <filterFile>] input.bb  > output.bed )

# Generate JACKIE.queryDB command using web form
  
For your convenience, we have webforms that can help you generate JACKIE.queryDB command on:

[One Copy hg38 CRISPR/SpCas9 JACKIEdb](https://cheng.bio/generate-jackiequerydb-command/?infilename=hg38PAM.1copy.offSiteCounts.wGCT.bed.gz)

[2+ Copies <50k cluster hg38 CRISPR/SpCas9 JACKIEdb](https://cheng.bio/generate-jackiequerydb-command/?multi=yes&infilename=hg38PAM.2PlusCopy.offSiteCounts.wGCT.within50kb.bed.gz)
  
[One Copy mm10 CRISPR/SpCas9 JACKIEdb](https://cheng.bio/generate-jackiequerydb-command/?infilename=mm10PAM.1copy.offSiteCounts.wGCT.bed.gz)

[2+ Copies <50k cluster  mm10 CRISPR/SpCas9 JACKIEdb](https://cheng.bio/generate-jackiequerydb-command/?multi=yes&infilename=mm10PAM.2PlusCopy.offSiteCounts.wGCT.within50kb.bed.gz)  
  
# Basic operations with examples

Our example bigbed https://albertcheng.info/jackie_downloads/hg38PAM.1copy.offSiteCounts.wGCT.bb which is an annotation of all 1 copy CRISPR/SpCas9 sites with their off-target site counts generated by JACKIE2 (https://cheng.bio/JACKIE2)

It has the following structure (in autosql)
```
table hg38OneCopyPAMOffSiteCounts
"One-copy CRISPR SpCas9 NGG sites and off-target counts up to 3 mismatches"
(
string  chrom;		"Reference sequence chromosome or scaffold"
uint    chromStart;	"Start position of feature on chromosome"
uint    chromEnd;	"End position of feature on chromosome"
string  name;		"Name of gene"
uint    score;		"Score"
char[1] strand;		"+ or - for strand"
uint    thickStart;	"Coding region start"
uint    thickEnd;	"Coding region end"
uint  	reserved;	"Color"
uint    n0mismatches;  "Number of exact matched sites"
uint    n1mismatches;  "Number of 1-mismatch sites"
uint    n2mismatches;  "Number of 2-mismatch sites"
uint    n3mismatches;  "Number of 3-mismatch sites"
uint    totalOffSites; "Total number of 1,2,3-mismatch sites"
string  offSiteCounts; "String representation of offsite counts separated by slashes"
string  spacerSeq;  "Spacer sequence of gRNA"
uint    percentGC;  "Percent GC of spacer sequence"
uint    longestTandemT; "Longest run of T"
)
```
or in a table

Column# | Name | Description
--- | --- | --- 
1 | chrom | Reference sequence chromosome or scaffold
2 | chromStart | Start position of feature on chromosome
3 | chromEnd | End position of feature on chromosome
4 | name | Name of gene
5 | score | Score
6 | strand | + or - for strand
7 | thickStart | Coding region start
8 | thickEnd | Coding region end
9 | reserved | Color
10 | n0mismatches | Number of exact matched sites
11 | n1mismatches | Number of 1-mismatch sites
12 | n2mismatches | Number of 2-mismatch sites
13 | n3mismatches | Number of 3-mismatch sites
14 | totalOffSites | Total number of 1,2,3-mismatch sites
15 | offSiteCounts | String representation of offsite counts separated by slashes
16 | spacerSeq | Spacer sequence of gRNA
17 | percentGC | Percent GC of spacer sequence
18 | longestTandemT | Longest run of T

An example item (a 1 copy gRNA with 2 1-mismatch and 1 3-mismatch off-targets)
```
chr1	54472	54492	8124609316492752457.1/CCCGGTAGAGTATTTGAAGA	1	-	54472	54492	0,0,0	1	2	0	1	3	1/2/0/1	CCCGGTAGAGTATTTGAAGA	45	3
```

The genomic regions of interest are supplied in a txt file and supplied as an ```-filter=filter.txt``` option to the program
The filter.txt file should contains each line ```chr```tab```start```tab```end```, e.g.,
```
chr1	1	500000
chr2	1	1000000
```
Example: To run with filter file:
```
bedBedToBedPlus -filter=filter.txt https://albertcheng.info/jackie_downloads/hg38PAM.1copy.offSiteCounts.wGCT.bb 
```
The filter.txt file can also contain the input.bb filename by adding :input.bb, and in command line only run with ```bedBedToBedPlus -filter=filter.txt```
```
:https://albertcheng.info/jackie_downloads/hg38PAM.1copy.offSiteCounts.wGCT.bb
chr1	1	500000
chr2	1	1000000
```
# Filtering
The filter.txt can contain ```$column_number``` tab ```min``` tab ```max``` to filter items with column integer values within min and max inclusive. tab delimited fields 4+ are ignored and can be used for commenting, ```#COMMENTS```. 
```
:https://albertcheng.info/jackie_downloads/hg38PAM.1copy.offSiteCounts.wGCT.bb
chr1	1	500000
chr2	1	1000000
$17	40	60	#filter for column 17 in range [40,60]
$14	0	4	#filter for column 14 in range [0,4]
$18	0	5	#filter for column 18 in range [0,5]
```
output (first top 10 lines):
```
chr1	10795	10815	8415868856704206347.1/GCAGACACATGCTAGCGCGT	1	+	10795	10815	0,0,0	1	0	0	0	0	1/0/0/0	GCAGACACATGCTAGCGCGT	60	1
chr1	47090	47110	8561696403163264064.1/AACTGTTTCTCTAGATCTGG	1	-	47090	47110	0,0,0	1	2	0	2	4	1/2/0/2	AACTGTTTCTCTAGATCTGG	40	3
chr1	51899	51919	8564449090403309194.1/TCTGCATGTACCCTTAGGGG	1	-	51899	51919	0,0,0	1	0	1	3	4	1/0/1/3	TCTGCATGTACCCTTAGGGG	55	2
chr1	52036	52056	8401791121874748634.1/TGGTAAAGAAGCACTCCGTT	1	+	52036	52056	0,0,0	1	0	0	4	4	1/0/0/4	TGGTAAAGAAGCACTCCGTT	45	2
chr1	54472	54492	8124609316492752457.1/CCCGGTAGAGTATTTGAAGA	1	-	54472	54492	0,0,0	1	2	0	1	3	1/2/0/1	CCCGGTAGAGTATTTGAAGA	45	3
chr1	54476	54496	8255500066069159937.1/CAAATACTCTACCGGGCTTC	1	+	54476	54496	0,0,0	1	2	0	2	4	1/2/0/2	CAAATACTCTACCGGGCTTC	50	2
chr1	63667	63687	8508183789487830611.1/GTCGGCTCACATCACCGTAG	1	+	63667	63687	0,0,0	1	2	0	1	3	1/2/0/1	GTCGGCTCACATCACCGTAG	60	1
chr1	69533	69553	8077523290482021064.1/ACGCCAACTGGCTCACCGAA	1	-	69533	69553	0,0,0	1	2	0	2	4	1/2/0/2	ACGCCAACTGGCTCACCGAA	60	1
chr1	69545	69565	8214896123759040152.1/AGTGCACGGCAAACGCCAAC	1	-	69545	69565	0,0,0	1	2	0	2	4	1/2/0/2	AGTGCACGGCAAACGCCAAC	60	1
chr1	263706	263726	8359534363805562568.1/ACGGGTGTTTTGGATAGAAT	1	-	263706	263726	0,0,0	1	0	0	2	2	1/0/0/2	ACGGGTGTTTTGGATAGAAT	40	4
(cropped)


```
# Selecting best items
The filter.txt can direct the program to only print out the best N items according to a sort lists, using
```!BEST```tab```number_of_items``` to specify the number of items to print per region
```!MIN```tab```column_number``` to minimize value in the specified column (i.e., sort ascending)
```!MAX```tab```column_number``` to maxmize value in the specified column (i.e., sort descending)

```
:https://albertcheng.info/jackie_downloads/hg38PAM.1copy.offSiteCounts.wGCT.bb
chr1	1	500000
chr2	1	1000000
$17	40	60	#filter for column 17 in range [40,60]
$14	0	4	#filter for column 14 in range [0,4]
$18	0	5	#filter for column 18 in range [0,5]
!BEST	4	#print best 4 items from each region, according to the list of sorting belong
!MIN	14	#minimize column 14 first (ascending)
!MAX	17	#then maximize column 17 (descending)
```
output
```
chr1	10795	10815	8415868856704206347.1/GCAGACACATGCTAGCGCGT	1	+	10795	10815	0,0,0	1	0	0	0	0	1/0/0/0	GCAGACACATGCTAGCGCGT	60	1
chr1	263706	263726	8359534363805562568.1/ACGGGTGTTTTGGATAGAAT	1	-	263706	263726	0,0,0	1	0	0	2	2	1/0/0/2	ACGGGTGTTTTGGATAGAAT	40	4
chr1	63667	63687	8508183789487830611.1/GTCGGCTCACATCACCGTAG	1	+	63667	63687	0,0,0	1	2	0	1	3	1/2/0/1	GTCGGCTCACATCACCGTAG	60	1
chr1	54472	54492	8124609316492752457.1/CCCGGTAGAGTATTTGAAGA	1	-	54472	54492	0,0,0	1	2	0	1	3	1/2/0/1	CCCGGTAGAGTATTTGAAGA	45	3
chr2	43144	43164	8233193175387283992.1/AGACTTACTGCCCGGCTACC	1	-	43144	43164	0,0,0	1	0	0	0	0	1/0/0/0	AGACTTACTGCCCGGCTACC	60	2
chr2	46239	46259	8528243356817207939.1/GATCGCGCCAAGTATGTGCG	1	-	46239	46259	0,0,0	1	0	0	0	0	1/0/0/0	GATCGCGCCAAGTATGTGCG	60	1
chr2	46241	46261	8546153689510617305.1/CGGATCGCGCCAAGTATGTG	1	-	46241	46261	0,0,0	1	0	0	0	0	1/0/0/0	CGGATCGCGCCAAGTATGTG	60	1
chr2	72890	72910	8539156629285802011.1/GGAACCAATACGGGGCCATG	1	-	72890	72910	0,0,0	1	0	0	0	0	1/0/0/0	GGAACCAATACGGGGCCATG	60	1

```

# Building JACKIE.queryDB
Download kent code: https://github.com/ucscGenomeBrowser/kent and follow build instructions

To build this, place this folder under kent source tree under folder kent/src/utils/ (i.e., make this folder kent/src/utils/JACKIE.queryDB) , then run ```make``` on kent/src level first, then go into JACKIE.queryDB and run ```make```.

